Fixes to Doc Build and Deployment

awholegunch opened the pull request:
These changes fix:
1. Errors in the deployment script, which created paths with two consecutive slashes in them ('//')
2. Makes major changes to the build scripts for docs, including:

- Changes destination of doc build from "build" to "target", matching maven conventions;
- Removes old functions no longer being used;
- Moves code between scripts so that common file has only code actually shared between build scripts;
- Removes colored log output on non-OS X machines;
- Changes the ``build.sh`` API to have an optimized ``all`` that shares a single Javadocs invocation between the Web and Github versions; and
- Cleans up errors and inconsistencies in internal naming and coding.

With these changes, a run on Bamboo can produce artifacts, and display them, without needing a staging server task. The other tasks in the Bamboo plan have been simplified, as only the CDAP repository is required .

Demonstration Runs:

- [Docs-only run, no Javadocs, ZIP and HTML artifacts](http://builds.cask.co/browse/CDAP-DOB1-2/artifact)
- [Docs-all run, Javadocs, ZIPs and HTML artifacts](http://builds.cask.co/browse/CDAP-DOB1-3/artifact)

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -14,19 +14,12 @@
-# Targets for both a limited and complete set of javadocs
-# Targets not included in usage are intended for internal usage by script
+# Common code for Build script for docs
+# Not called directly; included in either the main build.sh or the individual manual's build.sh
```
This isn't visible in the PR, but you shouldn't have a shebang header in a shell file that isn't meant to be executed.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -75,32 +69,24 @@ else
-if [ "x${3}" == "x" ]; then
-  TEST_INCLUDES="${TEST_INCLUDES_REMOTE}"
+
+if [[ "${OSTYPE}" == "darwin"* ]]; then
```
This should be controlled by a flag, not the operating system where the code is being run. Don't assume that there may never be Darwin-based build hosts (we currently have one).

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -127,9 +113,9 @@ function usage() {
   PROJECT_PATH=`pwd`
   echo "Build script for '${PROJECT_CAPS}' docs"
-  echo "Usage: ${SCRIPT} < option > [source test_includes]"
+  echo "Usage: ${SCRIPT} <option> [source]"
```
How about "action" since it's not really an option (or optional) and it really describes the action you want the script to take.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -140,25 +126,39 @@ function usage() {
+  else
+    echo "${1}${2}"
+  fi
+}
```
The conditional here is redundant. The output would be the same with either echo. Also, isn't enabling backslash escapes non-safe if you don't know the input? You would need to check and sanitize the input, or ensure that no input has backslashes in it.

awholegunch replied:
Fixed. Color is now set by environment variable rather than OS, and no longer uses the '-e' flag in echo statements.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -166,40 +166,29 @@ function build_docs_google() {
-function copy_javadocs_sdk() {
-  cd ${BUILD_PATH}/${HTML}
+function copy_javadocs_api() {
+  cd ${TARGET_PATH}/${HTML}
```
Don't change working directory. It's unnecessary.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -166,40 +166,29 @@ function build_docs_google() {
+  cd ${TARGET_PATH}/${HTML}
   rm -rf ${JAVADOCS}
-  cp -r ${SDK_JAVADOCS} .
+  cp -r ${API_JAVADOCS} .
```
`cp -r ${API_JAVADOCS} ${TARGET_PATH}/${HTML}`

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -216,18 +205,18 @@ function build_license_pdfs() {
 function copy_license_pdfs() {
-  cd ${BUILD_PATH}/${HTML}/${LICENSES}
+  cd ${TARGET_PATH}/${HTML}/${LICENSES}
   cp ${SCRIPT_PATH}/${LICENSES_PDF}/* .
```
Same as above. No need to `cd` only to `cp`.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -236,12 +225,6 @@ function make_zip() {
   zip -qr ${ZIP_DIR_NAME}.zip ${PROJECT_VERSION}/* --exclude .DS_Store
```
Why are we using a license-encumbered format for this artifact? This is an additional dependency that not all systems will have, just for building docs. It just *happens* that our build hosts all have a `zip` binary. If we're going to keep this, as is, then we need to add a capability to our build agents on `zip` and have the builds depend on agents with this capability. Hidden, undocumented dependencies are the devil.

awholegunch replied:
Why? Because that's what I was asked to do. People requested a zip file of the documentation.

awholegunch replied:
What would you suggest as a replacement? ``tar.gz``? That might not be acceptable on Windows.

wolf31o2 replied:
Ahh, so we're distributing the zip, not just using it, internally. Got it. That's a good enough reason. ZIP as a format is far more widespread and accessible.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -257,21 +240,31 @@ function build_web() {
+
 function set_mvn_environment() {
+  cd ${PROJECT_PATH}
   if [ "$(uname)" == "Darwin" ]; then
```
Be consistent. Use `${OSTYPE}` here and don't shell out to `uname` for this.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -279,12 +272,21 @@ function check_includes() {
 }
 
 function download_includes() {
+  local includes_dir=${1}
```
Why add this line? It's unused in this function.

awholegunch replied:
Because this function is a template to be over-shadowed in the inner manuals that use this. I was thinking I could then drop the comment, as this documents what $1 is.

awholegunch replied:
Fixed: removed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -279,12 +272,21 @@ function check_includes() {
-function test_includes () {
+function test_includes_directory() {
+  local includes_dir=${1}
+  if [ ! -d "${includes_dir}" ]; then
```
What if it's a file? This will pass and your mkdir below will fail. If this is in a transitive directory structure created by the run, it should be safe for you to `rm -rf ${includes_dir} && mkdir ${includes_dir}`

awholegunch replied:
Removed function, as it isn't needed. The directory is always created right before this step, so it is a redundant check.

wolf31o2 replied:
Sweet. :smile:

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -511,13 +490,10 @@ function run_command() {
-    depends )           build_dependencies;;
-    sdk )               build_sdk;;
     version )           display_version;;
     * )                 usage;;
```
This whole section could be simpler if you had a 1:1 command:function naming... perhaps with `s/-/_/` applied...

fex.
```
check-includes|docs|copy-javadocs|copy-license-pdfs) ${1/-/_};;
*) usage;;
```

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -257,21 +240,31 @@ function build_web() {
 }
 
 function check_includes() {
   if [ "${CHECK_INCLUDES}" == "${TRUE}" ]; then
```
Why compare string to string here? Why even have TRUE as a constant? Why not just evaluate CHECK_INCLUDES as a boolean?

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -279,12 +272,21 @@ function check_includes() {
+  local includes_dir=${1}
+  if [ ! -d "${includes_dir}" ]; then
+    echo "Creating Includes Directory: ${includes_dir}"
+    mkdir ${includes_dir}
```
Also, you're guarding the test against spaces in the path, by putting it in quotes, but not doing the same here. If you know there won't be spaces in the path, drop the quotes in the test. Otherwise, add quotes here.

awholegunch replied:
Deleted the function.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -30,6 +30,34 @@ ARG_1="${1}"
 
+function usage() {
+  cd ${PROJECT_PATH}
+  PROJECT_PATH=`pwd`
```
Pretty sure this is wrong. If it isn't, you should use a different variable name if this can have different values at different times.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -30,6 +30,34 @@ ARG_1="${1}"
 ARG_2="${2}"
 ARG_3="${3}"
```
Positional parameters are positional parameters. There's no need to wrap them in quotes, even if they includes spaces and are quoted in the input.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -30,6 +30,34 @@ ARG_1="${1}"
+function usage() {
+  cd ${PROJECT_PATH}
+  PROJECT_PATH=`pwd`
+  echo "Build script for '${PROJECT_CAPS}' docs"
```
Do we need two constants for the project, when this is part of CDAP?

awholegunch replied:
Similar versions of this code are used in Coopr and Tigon, hence the usage.

wolf31o2 replied:
Fair enough. We'll keep it as-is until we revamp for the new build after 3.2 is released, then we'll port that work to the other projects.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,227 +66,197 @@ function set_project_path() {
-}
-
 function run_command() {
   case "${1}" in
```
No need to quote. Best practice is to be conservative with quotes and only use them where necessary.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,227 +66,197 @@ function set_project_path() {
+  run_command docs-web-part ${ARG_2} 
+  _restore_github_zip
+  echo "--------------------------------------------------------"
+  _bell "Completed \"build all\""
```
Single quotes on the outside to avoid escaping quotes on the inside.

awholegunch replied:
That didn't always work: 
- ``'Building "${i}", target "${1}"...'`` didn't work, while 
- ``"Building \"${i}\", target \"${1}\"..."`` does...

wolf31o2 replied:
Correct. When there's variables, you need to use double quotes for them to be interpreted. Simple strings like these can be single-quoted... or is that not working after being passed to the function?

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -30,6 +30,34 @@ ARG_1="${1}"
+  echo "  with"
+  echo "    source    Path to ${PROJECT} source, if not '${PROJECT_PATH}'"
+  echo "              Path is relative to '${SCRIPT_PATH}/../..'"
+  echo ""
```
`echo` gives the same as `echo ""`

awholegunch replied:
I thought for sure I ran into a problem with using ``echo`` versus ``echo ""``, but...

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,227 +66,197 @@ function set_project_path() {
+    build_docs_outer_level ${source_path} 
+    copy_docs_inner_level
+  fi
+  local warnings="$?"
```
No quotes.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,227 +66,197 @@ function set_project_path() {
+    copy_docs_inner_level
+  fi
+  local warnings="$?"
+  if [ "${warnings}" != "0" ]; then
```
Compare int to int. No need to convert to string.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,227 +66,197 @@ function set_project_path() {
+
+function build_javadocs_api() {
+  set_mvn_environment
+  MAVEN_OPTS="-Xmx1024m -XX:MaxPermSize=128m" mvn clean install -P examples,templates,release -DskipTests -Dgpg.skip=true && mvn clean site -DskipTests -P templates -DisOffline=false
```
What are we doing here, especially given we're running `clean` in both Maven invocations?

awholegunch replied:
set_mvn_environment:
- Sets the directory that the command is invoked from
- On OS X (darwin), uses a specific version of Java
- Checks to see if the BUILD.rst file has changed: in that case, the command may need updating, as that is the source of this command.

wolf31o2 replied:
Sorry, I meant the actual two `mvn` calls. Both have clean in them. Are we simply trying to fill the Maven cache with the first? Anything built will otherwise be removed by the clean in the next `mvn` command, so I am not sure why we're doing them both this way.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,227 +66,197 @@ function set_project_path() {
-function build_javadocs() {
-  # Uses function from common
-  build_javadocs_api
+  return $?
```
This will return 0, always. The previous command is `echo`... maybe you meant to save the return value of a previous command and return with that?

awholegunch replied:
Yes. I added the echo statements and forgot to capture the value. Thanks.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,227 +66,197 @@ function set_project_path() {
+  echo "Building GitHub Docs"
+  echo "--------------------------------------------------------"
+  echo ""
+  _build_docs "build-github" ${GOOGLE_ANALYTICS_GITHUB} ${GITHUB} ${FALSE}
```
No quotes.

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,227 +66,197 @@ function set_project_path() {
+  echo "========================================================"
+  echo ""
+  cd ${SCRIPT_PATH}
+  TARGET_TEMP=$(mktemp -d /tmp/cdap-docs.XXXX)
```
Why are you writing outside of the build area? This breaks any kind of sandboxing of the build. Use a relative path for this.

awholegunch replied:
I was writing outside the build area because the build area was getting blown away between invocations. But I think that with the revisions, this may no longer be required (the stashing between invocations.) I will test and eliminate this somehow.

wolf31o2 replied:
Ahh, yeah. The other solution would be to build under build and save the artifacts under target. Then, you can wipe out build and still keep previous artifacts. This is how `mvn` does things, unless you run clean, which will wipe target, also.

awholegunch replied:
I will need a tmp area, as the stashing between invocations is still required. But I will modify the code so this happens within the build area.

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,227 +66,197 @@ function set_project_path() {
+  echo "========================================================"
+  echo ""
+  mv ${TARGET_TEMP}/*.zip ${SCRIPT_PATH}/${TARGET}
+  rm -rf ${TARGET_TEMP}
```
Rather than doing cleanup inside functions where a previous command could fail, it would make sense to instead trap signals and have an exit handler which performs cleanup.

```
trap trap_cleanup SIGHUP SIGQUIT SIGINT SIGTERM SIGKILL
function trap_cleanup() { cleanup; exit 1; }
function cleanup() {
  # perform cleanups here
}
```

awholegunch replied:
This will probably just go away...

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -275,53 +273,145 @@ function _build_docs() {
 
-function build_zip() {
+function _build_zip() {
   cd $SCRIPT_PATH
```
Missing braces

awholegunch replied:
Fixed.

awholegunch commented:
Now on to testing...

awholegunch commented:
I have updated this, fixed all the issues mentioned above, and in particular have removed use of ``/tmp``, and so everything happens inside ``target`` directories. Removed changing of the current directory wherever possible.

Tested both locally and on Bamboo.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -14,19 +12,16 @@
+# Common code for Build script for docs
+# Not called directly; included in either the main build.sh or the individual manual's build.sh
+
+# Optional Parameters (passed via Bamboo env variable or exported in shell):
```
Remove Bamboo

awholegunch replied:
Fixed (two places.)

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -124,82 +112,62 @@ BUILD_RST_HASH="f54ae74bb72f9ad894766b6c0bd2d2df"
   echo "  with"
-  echo "    source         Path to $PROJECT source for javadocs, if not $PROJECT_PATH"
-  echo "    test_includes  local, remote or neither (default: remote); must specify source if used"
+  echo "    source               Path to $PROJECT source for javadocs, if not '$PROJECT_PATH'"
```
Add braces

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/_common/common-build.sh:
```diff
@@ -257,21 +218,31 @@ function build_web() {
 function check_includes() {
-  if [ "${CHECK_INCLUDES}" == "${TRUE}" ]; then
-    echo "Downloading and checking includes."
+  if [ !"${CHECK_INCLUDES}" ]; then
```
It looks like this changed the meaning of the check. Can you verify that this is correct?

awholegunch replied:
Good catch. I went back to the old check, and checked all usages of this variable.

wolf31o2 commented:
Couple minor comments and one question... otherwise, :+1: LGTM

awholegunch commented:
Running another test on Bamboo to check latest version...

awholegunch commented:
When running additional tests, I found that messages from the Javadoc portion of the build were not being retained and replayed at the end, and that the return code was not being passed to Bamboo successfully. This has been fixed in the last two commits.

awholegunch commented:
The build is now successfully failing!

wolf31o2 commented on cdap-docs/build.sh:
```diff
@@ -38,321 +72,379 @@ function set_project_path() {
+  fi
+  display_version
+  display_messages_file
+  warnings=$?
```
Return code from `display_messages_file`?

awholegunch replied:
Yes...``display_messages_file`` determines the return code based on the contents of the message file. Once all the messages have been displayed at what will become the end of the log, it then returns the code and Bamboo can decide what to do with the results. This way, when you go to the Bamboo page, you should be able to see at a glance that there are error messages to be reviewed, and which manuals to poke into.

wolf31o2 commented:
One minor comment... otherwise, :+1: LGTM
